### PR TITLE
[Merged by Bors] - feat: port Data.Matrix.Notation

### PR DESCRIPTION
**Pull request by eric-wieser**
This PR also fixes the doc on `Data.Fin.VecNotation` and adds the unexpander for the `![x, y, ...]` notation.

---

- [x] depends on: #3215
- [x] depends on: #3430
- [x] depends on: https://github.com/leanprover-community/mathport/pull/235
- [x] depends on: #3485

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#3215
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#3215~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by Komyyy** [comment]
I think the cause of the error is that `em'` is reduced to `toTypeExpr m'` and Qq tries to antiquote `toTypeExpr m'` nevertheless it doesn't have `Q(_)` type.
This problem is solved by replace `let` with `have`.

**Comment by eric-wieser** [comment]
Let's wait for mathport support before considering merging this.

File: test/matrix.lean [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+set_option pp.universes true
+set_option pp.all true
+
+meta def get_dims (e : pexpr) : tactic (Expr × Expr) :=
+do
+  elem_t ← tactic.mk_meta_var (expr.sort level.zero.succ),
+  e ← tactic.to_expr ``(%%e : matrix _ _ %%elem_t) tt ff,
+  t ← tactic.infer_type e,
+  `(Matrix.{0 0 0} (Fin %%m) (Fin %%n) %%elem_t) ← tactic.infer_type e,
+  return (m, n)
```
Comment (eric-wieser) [inline test/matrix.lean:30]: This ought to be straightforward to port, though I haven't tried yet.

File: test/matrix.lean [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+example {a b c d e : α} : ![a, b, c, d, e] 7 = c := by simp
+example {a b c d e : α} : ![a, b, c, d, e] 8 = d := by simp
+example {a b c d e : α} : ![a, b, c, d, e] 9 = e := by simp
+example {a b c d e : α} : ![a, b, c, d, e] 123 = d := by simp
+example {a b c d e : α} : ![a, b, c, d, e] 123456789 = e := by simp
+
+example {a b c d e f g h : α} : ![a, b, c, d, e, f, g, h] 5 = f := by simp
+example {a b c d e f g h : α} : ![a, b, c, d, e, f, g, h] 7 = h := by simp
+example {a b c d e f g h : α} : ![a, b, c, d, e, f, g, h] 37 = f := by simp
+example {a b c d e f g h : α} : ![a, b, c, d, e, f, g, h] 99 = d := by simp
```
Comment (eric-wieser) [inline test/matrix.lean:97]: If these tests fail we can just comment them out; they're not related to this PR

File: test/matrix.lean [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  let n ← instantiateMVars n
+  if m.hasMVar || n.hasMVar then
+    throwError "indices have metavariables"
+  return (m, n)
+
+open Lean Elab in
+def run (x : TermElabM α) : Command.CommandElabM α := Command.liftTermElabM x
+
+-- we test equality of expressions here to ensure that we have `2` and not `1.succ` in the type
+run_cmd run do let d ← getDims ``(!![]);        guard $ d == (q(0), q(0))
```
Comment (kmill) [inline test/matrix.lean:47]: I'm not completely sure what double backquote is, but I'm under the impression it's a version of quoting where it checks that every identifier refers to something. Single backquotes are probably more appropriate here.
```suggestion
run_cmd run do let d ← getDims `(!![]);        guard $ d == (q(0), q(0))
```
(Lean 3 was pexprs vs exprs, right?)

**Comment by digama0** [comment]
> Let's wait for mathport support before considering merging this.

I think you want more than mathport support for this. Is the notation itself implemented yet? Otherwise I don't see how you can actually use it in the file.

File: Mathlib/Data/Matrix/Notation.lean [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+end toExpr
+
+section Parser
+open Lean Elab Term Macro TSyntax
+
+syntax "!![" sepBy1(sepBy1(term, ",", ", ", allowTrailingSep),
+              ";", "; ", allowTrailingSep) "]" : term
+syntax "!![" ";"+ "]" : term
+syntax "!![" ","+ "]" : term
+syntax "!![" "]" : term
```
Comment (digama0) [inline Mathlib/Data/Matrix/Notation.lean:87]: This syntax differs from #3429, so it will need another mathport bump if not reverted

**Comment by Parcly-Taxel** [comment]
> Let's wait for mathport support before considering merging this.

Well… https://github.com/leanprover-community/mathport/pull/235 _is_ ready now.

File: Mathlib/Data/Matrix/Notation.lean [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+This ensures `simp` works with entries only when (some) entries are already given.
+In other words, this notation will only appear in the output of `simp` if it
+already appears in the input.
+
+## Notations
+
+This file provide notation `!![a, b; c, d]` for matrices, which corresponds to
+`Matrix.of ![![a, b], ![c, d]]`.
+Note that in lean 4 the pretty-printer will not show `!!` notation, instead showing the version
+with `of ![![...]]`.
```
Comment (Komyyy) [inline Mathlib/Data/Matrix/Notation.lean:41]: Lean didn't delaborate `![x, y, ...]` notations, so this was wrong.
So, I added unexpanders.

File: Mathlib/Data/Matrix/Notation.lean (hunk 4 above)
Comment (Komyyy) [reply Mathlib/Data/Matrix/Notation.lean]: I synchronized syntaxes with `Mathport.Syntax`.

**Comment by eric-wieser** [comment]
> > Let's wait for mathport support before considering merging this.
> 
> Well… [leanprover-community/mathport#235](https://github.com/leanprover-community/mathport/pull/235) _is_ ready now.

Great! What I mean is that we should wait for that to be merged and mathlib3port to be regenerated. Until then, all the downstream files in mathport will have holes in where this notation is used; so we don't want people trying to port them yet.

**Comment by semorrison** [comment]
I merged the mathport PR. Hopefully tomorrow we will see what happens in mathlib3port.

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#3215~~
* ~~leanprover-community/mathlib4#3430~~
* ~~leanprover-community/mathport#235~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: test/matrix.lean (hunk 3 above)
Comment (eric-wieser) [reply test/matrix.lean]: Done

File: Mathlib/Data/Matrix/Notation.lean (hunk 5 above)
Comment (eric-wieser) [reply Mathlib/Data/Matrix/Notation.lean:41]: It doesn't look like you pushed the unexpander?

File: Mathlib/Data/Matrix/Notation.lean [hunk 6]
```diff
@@ -11,157 +11,129 @@ Authors: Anne Baanen, Eric Wieser
...
 
 ## Notations
 
 This file provide notation `!![a, b; c, d]` for matrices, which corresponds to
-`matrix.of ![![a, b], ![c, d]]`.
-A parser for `a, b; c, d`-style strings is provided as `matrix.entry_parser`, while
-`matrix.notation` provides the hook for the `!!` notation.
-Note that in lean 3 the pretty-printer will not show `!!` notation, instead showing the version
+`Matrix.of ![![a, b], ![c, d]]`.
+Note that in lean 4 the pretty-printer will not show `!!` notation, instead showing the version
```
Comment (eric-wieser) [inline Mathlib/Data/Matrix/Notation.lean:40]: We should add a TODO here to implement the unexpander; we didn't have an option in Lean 3, hence the original comment.

File: Mathlib/Data/Matrix/Notation.lean [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+syntax (name := matrixNotation0xC) "!![" ","+ "]" : term
+
+macro_rules
+  | `(!![$[$[$rows],*];*]) => do
+    let m := rows.size
+    let n := rows[0]!.size
+    let rowVecs ← rows.mapM fun row : Array Term => do
+      unless row.size = n do
+        Macro.throwErrorAt (mkNullNode row)
+          s!"Rows must be of equal length; this row has {row.size} items, the previous rows have {n}"
```
Comment (eric-wieser) [inline Mathlib/Data/Matrix/Notation.lean:91]: What's the preferred way to line-wrap this?

Comment (digama0) [reply Mathlib/Data/Matrix/Notation.lean:91]: My trick for line-wrapping interpolated strings is to do e.g.
```
"My really really long {""
}line"
```

File: test/matrix.lean [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+-- we test equality of expressions here to ensure that we have `2` and not `1.succ` in the type
+#guard dims% !![]        is (0, 0)
+#guard dims% !![;]       is (1, 0)
+#guard dims% !![;;]      is (2, 0)
+#guard dims% !![,]       is (0, 1)
+#guard dims% !![,,]      is (0, 2)
+#guard dims% !![1]       is (1, 1)
+#guard dims% !![1,]      is (1, 1)
+#guard dims% !![1;]      is (1, 1)
+#guard dims% !![1,2;3,4] is (2, 2)
```
Comment (kmill) [inline test/matrix.lean:79]: @eric-wieser I just pushed a test framework that we can eventually generalize out of this file. The matrix-local part of it is this `dims%` elaborator.

File: test/matrix.lean [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+/-! Test that the dimensions are inferred correctly, even for empty matrices -/
+section dimensions
+
+-- set_option pp.universes true
+-- set_option pp.all true
+
+section elaborators
+open Lean Meta Elab Command
+
+/-- `dims% e` elaborates `e` as a Matrix and returns its dimensions as a `Nat × Nat`. -/
```
Comment (eric-wieser) [inline test/matrix.lean:30]: This isn't testing the right thing any more; the original was intended to test that the dimensions were syntactically `2` and not `succ (succ 0)` or similar.

Comment (eric-wieser) [reply test/matrix.lean:30]: Ah, perhaps the docstring is just wrong; looking more closely it looks like what you have is testing the right thing after all.

File: test/matrix.lean [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+  let m ← mkFreshExprMVar (mkConst ``Nat)
+  let n ← mkFreshExprMVar (mkConst ``Nat)
+  let matrix_t := mkAppN (mkConst ``Matrix [levelZero, levelZero, levelZero])
+                    #[mkApp (mkConst ``Fin) m, mkApp (mkConst ``Fin) n, elem_t]
+  let _ ← Term.elabTermEnsuringType e (some matrix_t)
+  let m ← instantiateMVars m
+  let n ← instantiateMVars n
+  mkAppM ``Prod.mk #[m, n]
+
+/-- `#guard e1 is e2` elaborates `e1` and `e2` and checks they are `BEq`. -/
```
Comment (eric-wieser) [inline test/matrix.lean:42]:
```suggestion
/-- `#guard e1 is e2` elaborates `e1` and `e2` and checks they are `BEq` as `Expr`s. -/
```

File: test/matrix.lean (hunk 9 above)
Comment (kmill) [reply test/matrix.lean:30]: No, it is. The `#guard` command is checking syntactic equality.

```
#guard dims% !![;;]      is (2, 0) -- succeeds
#guard dims% !![;;]      is (Nat.succ 1, 0) -- fails
```

File: test/matrix.lean (hunk 10 above)
Comment (kmill) [reply test/matrix.lean:42]: I'll fix this. It'll be a `#guard_expr` command that more similarly matches the `guard_expr` tactic.

Comment (kmill) [reply test/matrix.lean]: Ok, it's now using the syntax from `guard_expr`, so it looks like
```
#guard_expr dims% !![1,2;3,4] =ₛ (2, 2)
```
It handles defeq testing too with `=~`:
```
#guard_expr dims% !![1,2;3,4] =~ (Nat.succ 1, 2)
```
Supposedly `guard_expr` handles `=` for equality at reducible transparency, but I can't get it to parse.

Comment (digama0) [reply test/matrix.lean]: > Supposedly `guard_expr` handles `=` for equality at reducible transparency, but I can't get it to parse.

You are probably forgetting to use `x:term:51 " = " y:term` or else `x` will also consume the `=` token.

File: test/matrix.lean [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+  mkAppM ``Prod.mk #[m, n]
+
+/-- Check equality of two expressions.
+* `#guard_expr e = e'` checks that `e` and `e'` are defeq at reducible transparency.
+* `#guard_expr e =~ e'` checks that `e` and `e'` are defeq at default transparency.
+* `#guard_expr e =ₛ e'` checks that `e` and `e'` are syntactically equal.
+* `#guard_expr e =ₐ e'` checks that `e` and `e'` are alpha-equivalent.
+
+This is similar to the `guard_expr` tactic but also uses default instances and instantiates
+metavariables. -/
```
Comment (digama0) [inline test/matrix.lean:50]: I don't like this slight difference in behavior. These little things are exactly why there are so many variations on guard_expr in the first place; there were 3 or 4 different functions, each one adding another delta on the last. If you need this behavior just PR it to std and make things consistent.

Comment (kmill) [reply test/matrix.lean:50]: I don't either, and truth be told that's why I left a note. I think the `guard_expr` tactic should also do metavariable instantiation and also make sure to use `synthesizeSyntheticMVarsUsingDefault`. (Perhaps `MatchKind.isEq` should be responsible for instantiating metavariables?)

This also isn't a simple matter of making a PR to Std to make things consistent: Std has a `guard_expr` tactic elaborator, but this is a `guard_expr#` command elaborator. I'd be happy if someone made the PR to include the new elaborator and make things consistent (and ideally even factoring out the body of both elaborators). Maybe I will do this eventually, but it's not high priority for me, and I'm happy enough with the partial progress toward improving `test/matrix.lean`.

File: Mathlib/Data/Matrix/Notation.lean [hunk 12]
```diff
@@ -0,0 +1,499 @@
...
+open Lean Elab Term Macro TSyntax
+
+syntax (name := matrixNotation) "!![" sepBy1(term,+,?, ";", "; ", allowTrailingSep) "]" : term
+syntax (name := matrixNotationRx0) "!![" ";"* "]" : term
+syntax (name := matrixNotation0xC) "!![" ","+ "]" : term
+
+macro_rules
+  | `(!![$[$[$rows],*];*]) => do
+    let m := rows.size
+    let n := rows[0]!.size
```
Comment (gebner) [inline Mathlib/Data/Matrix/Notation.lean:87]:
```suggestion
    let n := if h : m > 0 then rows[0].size else 0
```

You should always avoid panicking when handling unexpected syntax.  In this case it's kind of okay, but in general the parser does error recovery by inserting `.missing` nodes.  And those will break all invariants you could assume.

And then there's of course:
```lean
def mkColumnVector (elems : Array Term) : Term :=
  Unhygienic.run `(!![$[$elems];*])
```

Comment (kmill) [reply Mathlib/Data/Matrix/Notation.lean]: Added a test to check that Gabriel's `mkColumnVector` works with the empty array.

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#3215~~
* ~~leanprover-community/mathlib4#3430~~
* ~~leanprover-community/mathport#235~~
* ~~leanprover-community/mathlib4#3485~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: test/matrix.lean (hunk 11 above)
Comment (semorrison) [reply test/matrix.lean:50]: Could `#guard_expr` at least be split out as a separate PR? It's fantastic to have, but I think less confusing if it arrives in a separate PR/commit. I'll promise to merge quickly. :-)

Comment (semorrison) [reply test/matrix.lean:50]: This PR is holding up progress on many fronts. Let's merge this and clean up `guard_expr` later.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**